File: src/main/java/org/mercurialftc/mercurialftc/scheduler/commands/SequentialCommandGroup.java
package org.mercurialftc.mercurialftc.scheduler.commands;

import org.jetbrains.annotations.Nullable;
import org.mercurialftc.mercurialftc.scheduler.OpModeEX;
import org.mercurialftc.mercurialftc.scheduler.Scheduler;
import org.mercurialftc.mercurialftc.scheduler.subsystems.SubsystemInterface;

import java.util.*;

@SuppressWarnings("unused")
public class SequentialCommandGroup implements CommandGroup {
	private final ArrayList<Command> commands;
	private final boolean interruptable;
	private final Set<SubsystemInterface> requiredSubsystems;
	private final Set<OpModeEX.OpModeEXRunStates> runStates;
	private int commandIndex;
	@Nullable
	private Command currentCommand;

	public SequentialCommandGroup() {
		this.requiredSubsystems = new HashSet<>();
		this.runStates = new HashSet<>(2);
		interruptable = true;
		this.commands = new ArrayList<>();
		commandIndex = -1;
	}

	private SequentialCommandGroup(ArrayList<Command> commands, Set<SubsystemInterface> requirements, Set<OpModeEX.OpModeEXRunStates> runStates, boolean interruptable) {
		this.requiredSubsystems = requirements;
		this.runStates = runStates;
		this.commands = commands;
		this.interruptable = interruptable;
		commandIndex = -1;
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new SequentialCommandGroup, with the added commands
	 */
	public SequentialCommandGroup addCommands(Command... commands) {
		return addCommands(Arrays.asList(commands));
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new SequentialCommandGroup, with the added commands
	 */
	public SequentialCommandGroup addCommands(Collection<Command> commands) {
		if (commandIndex != -1) {
			throw new IllegalStateException(
					"Commands cannot be added to a composition while it is running");
		}

		ArrayList<Command> newCommandList = new ArrayList<>(this.commands);
		newCommandList.addAll(commands);

		Set<SubsystemInterface> newRequirementSet = new HashSet<>(this.getRequiredSubsystems());
		boolean newInterruptable = this.interruptible();

		HashSet<OpModeEX.OpModeEXRunStates> newRunStates = new HashSet<>(2);

		for (Command command : commands) {
			newRequirementSet.addAll(command.getRequiredSubsystems());
			newInterruptable &= command.interruptible();
			newRunStates.addAll(command.getRunStates());
		}

		Scheduler.getSchedulerInstance().registerComposedCommands(commands);

		return new SequentialCommandGroup(
				newCommandList,
				newRequirementSet,
				newRunStates,
				newInterruptable
		);
	}

	@Override
	public final boolean interruptible() {
		return interruptable;
	}

	@Override
	public final void initialise() {
		if (commands.isEmpty()) {
			throw new RuntimeException("Attempted to run empty SequentialCommandGroup, SequentialCommandGroupRequires requires a minimum of 1 Command to be run");
		}
		commandIndex = 0;
		currentCommand = commands.get(commandIndex);
		if (currentCommand.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
			currentCommand.initialise();
		}
	}

	@Override
	public final void execute() {
		if (currentCommand == null) return;
		if (currentCommand.finished() || !currentCommand.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
			if (currentCommand.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
				currentCommand.end(false);
			}
			commandIndex++;
			if (commandIndex < commands.size()) {
				currentCommand = commands.get(commandIndex);
				if (currentCommand.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
					currentCommand.initialise();
				}
			} else {
				currentCommand = null;
				return;
			}
		}
		if (currentCommand.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState())) {
			currentCommand.execute();
		}
	}

	@Override
	public final void end(boolean interrupted) {
		commandIndex = -1;
		if (currentCommand == null || !currentCommand.getRunStates().contains(Scheduler.getSchedulerInstance().getRunState()))
			return;
		currentCommand.end(interrupted);
	}

	@Override
	public final boolean finished() {
		return commandIndex >= commands.size() || currentCommand == null;
	}

	@Override
	public final Set<SubsystemInterface> getRequiredSubsystems() {
		return requiredSubsystems;
	}

	@Override
	public final Set<OpModeEX.OpModeEXRunStates> getRunStates() {
		return runStates;
	}
}

File: Mercurial/src/main/kotlin/dev/frozenmilk/mercurial/commands/SequentialCommandGroup.kt
package dev.frozenmilk.mercurial.commands

import dev.frozenmilk.mercurial.Mercurial
import dev.frozenmilk.mercurial.RunStates
import dev.frozenmilk.mercurial.subsystems.Subsystem
import java.util.Arrays

class SequentialCommandGroup : CommandGroup {
	private val commands: ArrayList<Command>
	override val requiredSubsystems: Set<Subsystem>
	override val runStates: Set<RunStates>
	private var _interruptible: Boolean
	private var commandIndex: Int
	private var currentCommand: Command? = null

	constructor() {
		requiredSubsystems = HashSet()
		runStates = HashSet(2)
		commands = ArrayList()
		commandIndex = -1
		_interruptible = true
	}

	private constructor(commands: ArrayList<Command>, requirements: Set<Subsystem>, runStates: Set<RunStates>, interruptible: Boolean) {
		requiredSubsystems = requirements
		this.runStates = runStates
		this.commands = commands
		commandIndex = -1
		this._interruptible = interruptible
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new SequentialCommandGroup, with the added commands
	 */
	override fun addCommands(vararg commands: Command): SequentialCommandGroup {
		return addCommands(Arrays.asList(*commands))
	}

	/**
	 * non-mutating
	 *
	 * @param commands new commands to add
	 * @return a new SequentialCommandGroup, with the added commands
	 */
	override fun addCommands(commands: Collection<Command>): SequentialCommandGroup {
		check(commandIndex == -1) { "Commands cannot be added to a composition while it is running" }
		val newCommandList = ArrayList(this.commands)
		newCommandList.addAll(commands)
		val newRequirementSet = HashSet(requiredSubsystems)
		var newInterruptible = interruptible
		val newRunStates = HashSet<RunStates>(2)
		for (command in commands) {
			newRequirementSet.addAll(command.requiredSubsystems)
			newInterruptible = newInterruptible and command.interruptible
			newRunStates.addAll(command.runStates)
		}
		Mercurial.registerComposedCommands(commands)
		return SequentialCommandGroup(
				newCommandList,
				newRequirementSet,
				newRunStates,
				newInterruptible
		)
	}

	override val interruptible: Boolean
		get() {
			return _interruptible
		}

	override fun initialise() {
		if (commands.isEmpty()) {
			throw RuntimeException("Attempted to run empty SequentialCommandGroup, SequentialCommandGroupRequires requires a minimum of 1 Command to be run")
		}
		commandIndex = 0
		currentCommand = commands[commandIndex]
		if (currentCommand!!.runStates.contains(Mercurial.runState)) {
			currentCommand!!.initialise()
		}
	}

	override fun execute() {
		if (currentCommand == null) return
		if (currentCommand!!.finished() || !currentCommand!!.runStates.contains(Mercurial.runState)) {
			if (currentCommand!!.runStates.contains(Mercurial.runState)) {
				currentCommand!!.end(false)
			}
			commandIndex++
			if (commandIndex < commands.size) {
				currentCommand = commands[commandIndex]
				if (currentCommand!!.runStates.contains(Mercurial.runState)) {
					currentCommand!!.initialise()
				}
			} else {
				currentCommand = null
				return
			}
		}
		if (currentCommand!!.runStates.contains(Mercurial.runState)) {
			currentCommand!!.execute()
		}
	}

	override fun end(interrupted: Boolean) {
		commandIndex = -1
		if (currentCommand == null || !currentCommand!!.runStates.contains(Mercurial.runState)) return
		currentCommand!!.end(interrupted)
	}

	override fun finished(): Boolean {
		return commandIndex >= commands.size || currentCommand == null
	}
}
